### ISSUE TITLE
Prettier Format on Save Does not Honor .prettierrc.json

### ISSUE THREAD

**Issue by crawfordleeds** [zed-industries/zed#6017]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Following the configuration guide in the docs [here](https://docs.zed.dev/languages/javascript), I have found prettier on save to not honor the .prettierrc.json file in the root of the file. Whenever I run `prettier --write .` from the root of my project it seems to work. I can tell because we have `"tabWidth": 4` but saving the file within Zed is reverting back to the default `"tabWidth": 2`.

### Environment

Zed: v0.77.3 (stable)
OS: macOS 13.0.0
Memory: 32 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

**Comment by Nickersoft** [comment]
Can confirm this doesn't seem to be working and still needed to rely on the workaround.

@cezarneaga Not sure if you're still running into this, but it looks like you have a typo in your settings. `"external"` belongs under the `"formatter"` key, not `"format_on_save"`. `"format_on_save"` needs to be set to `"on"`.

**Comment by SomeoneToIgnore** [comment]
To get more information about it: does your project declare prettier in `package.json`? 
We have a "default" prettier that is shipped with Zed and gets a few settings out its settings, and the ones we fetch from package.json that do not get any settings:

https://github.com/zed-industries/zed/blob/347f68887f645ea004becfb43d178727eb9802fa/crates/prettier/src/prettier.rs#L291-L314

So seems like you're using the default one and hitting that `is_default()` codepath.
The problem though, is that we're supposed to detect prettier configs later via prettier js integration:
https://github.com/zed-industries/zed/blob/347f68887f645ea004becfb43d178727eb9802fa/crates/prettier/src/prettier_server.js#L172-L188

and this gets broken somehow?
There's a way to see the resulting configs via `debug: open language server logs` command:
<img width="1357" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/cb21004e-1f5c-4820-988b-a8c47eb8c0c5">
to debug it better, but at least we need to understand the set-up to proceed with that.

**Comment by andreasvirkus** [comment]
@Nickersoft it's a little unintuitive because the official guide's examples have `external` config nested under the `format_on_save` settings, so if you modify those, that's most likely how you'd end up with a `formatter.external` setting 

https://zed.dev/docs/javascript

**Comment by klapec** [comment]
I have a prettier config placed within a separate package in my monorepo and this fixed prettier in Zed for me:

```json
  "format_on_save": {
    "external": {
      "command": "node_modules/.bin/prettier",
      "arguments": [
        "--config",
        "packages/configs/prettier.config.mjs",
        "--stdin-filepath",
        "{buffer_path}"
      ]
    }
  }
```
(make sure `--config` with the correct path is set **before** `--stdin-filepath`)

**Comment by max-petrenko** [comment]
I'm also having the issue of Zed ignoring prettier config specified in the `.prettierrc` file, but with the abovementioned config enabled prettier stops working at all.

The config:
```json
"format_on_save": "on",
  "formatter": {
    "external": {
      "command": "node_modules/.bin/prettier",
      "arguments": [
        "--stdin-filepath",
        "{buffer_path}"
      ]
    }
  },
```

Logs:
```log
Caused by:
    No such file or directory (os error 2)
2024-06-27T21:09:18+03:00 [ERROR] crates/editor/src/editor.rs:9502: failed to format via external command "node_modules/.bin/prettier"

Caused by:
    No such file or directory (os error 2)
```


Am I doing something wrong with the config?

**Comment by tomoima525** [comment]
It's been closed already but I'll leave what worked on my end.
I have a TypeScript project with prettier + eslint set. 

```
{
  "languages": {
    "TypeScript": {
      "code_actions_on_format": {
        "source.fixAll.eslint": true,
        "source.organizeImports": true
      }
    }
  }
}
```

I followed the instruction in https://zed.dev/docs/languages/javascript?highlight=eslint#eslint but for TypeScript

@max-petrenko Do you have a prettier file in `node_modules/.bin/prettier` ? The error says that does not exist

**Comment by laralove143** [comment]
I don't understand why this was closed. The issue still happens on me (MacOS). I've tried both configs:

```json
{
  "languages": {
    "Markdown": {
      "format_on_save": "on",
      "formatter": "prettier"
    }
  }
}
```
and
```json
{
  "format_on_save": "on",
  "languages": {
    "Markdown": {
      "formatter": "prettier"
    }
  }
}
```

Calling the format editor command manually works fine.

**Comment by aretrace** [comment]
@laralove143 same here.
@SomeoneToIgnore could you reopen?

**Comment by Gregoor** [comment]
The VSCode Prettier extension resolves the appropriate Prettier version, even for monorepos and only falls back to the editor version of that:
https://github.com/prettier/prettier-vscode?tab=readme-ov-file#prettier-resolution

It would be great if Zed could behave the same way!

**Comment by momotofu** [comment]
> I have a prettier config placed within a separate package in my monorepo and this fixed prettier in Zed for me:
> 
>   "format_on_save": {
>     "external": {
>       "command": "node_modules/.bin/prettier",
>       "arguments": [
>         "--config",
>         "packages/configs/prettier.config.mjs", <<< Update this
>         "--stdin-filepath",
>         "{buffer_path}"
>       ]
>     }
>   }
> (make sure `--config` with the correct path is set **before** `--stdin-filepath`)

This worked for me. Except I changed the config path to `"--config", ".prettier/.prettierrc", "--stdin-filepath", "{buffer_path}"`.

I hope this helps. This can be really frustrating. The docs could be much clearer.

**Comment by abue-ammar** [comment]
> I don't understand why this was closed. The issue still happens on me (MacOS). I've tried both configs:
> 
> {
>   "languages": {
>     "Markdown": {
>       "format_on_save": "on",
>       "formatter": "prettier"
>     }
>   }
> }
> and
> 
> {
>   "format_on_save": "on",
>   "languages": {
>     "Markdown": {
>       "formatter": "prettier"
>     }
>   }
> }
> Calling the format editor command manually works fine.

if i put this, do i need to install prettier in node module? or npm i -g prettier?

**Comment by JoRyGu** [comment]
Thanks @klapec and @momotofu. This is also the solution that worked for me, except I've got the config nested in "formatter" instead of "format_on_save."

```json
"languages": {
    "TypeScript": {
      "formatter": {
        "external": {
          "command": "node_modules/.bin/prettier",
          "arguments": [
            "--config",
            ".prettierrc.js",
            "--stdin-filepath",
            "{buffer_path}"
          ]
        }
      }
    }
  }
```
@SomeoneToIgnore, I think a lot of the confusion here could be alleviated with a different approach to the documentation for how some of the default formatters work. It's pretty hard to parse and this solution doesn't seem to exist in the docs that I can find. The implementation itself doesn't appear to be broken.

**Comment by grrowl** [comment]
If anyone's working in a JS/TS monorepo and having trouble, you need to have

```
  "devDependencies": {
    "prettier": "*"
  }
```

defined in each workspace package's `package.json` for Zed to allow it. Prettier will still use the `.prettierrc` from your monorepo root.
